[ fix #3564 ] Being the default is not the same as being set to the default

gallais opened the pull request:
Distinguishing between an option currently being the default and
it being set to a value which happens to mean the same as the default.

My end goal is for flags to get a semantics *as a set* rather than a list.
A different meaning when options are ordered differently is too unexpected.

* [x] Update documentation to reflect the current state of affairs

gallais committed:
[ re #3564 ] Playing with ternary options

Distinguishing between an option currently being the default and
it being set to a value which happens to mean the same as the default.

gallais committed:
[ fix ] bump version number

I was decoding True to Off & thus getting REALLY confusing errors...

gallais committed:
[ fix #3564 ] adding tests

asr commented:
@gallais, sorry for the white-space issues in `CHANGELOG` reported by Travis. Please note there are other issues in the suite of failing tests.

gallais committed:
[ fix ] Let unsafePragmaOptions raise exceptions

gallais committed:
[ refactor ] Utils.WithDefault, introduced Utils.TypeLits

gallais committed:
[ more ] withoutK is now a WithDefault 'False

gallais committed:
[ cosmetic ] spacing, clarifying comment in test case

[ci skip]

gallais committed:
[ refactor ] WithDefault.hs

After my experiment generalizing this pattern to any type [1] which
we can't use at the moment because it requires TypeInType (introduced
in ghc 8.0.1), I have decided to change the representation.

Now it's clearer that WithDefault is a fancy Maybe.

[1] https://github.com/gallais/potpourri/blob/41f9e06fc79a8d41cb906e1f97fa3d85b58a7ffb/haskell/defaulting/Default.hs

fredrikNordvallForsberg reviewed (approved):
Looks good to me, but should probably update the user manual and CHANGELOG. For the user manual, I think the note on how to turn on `--sized-types` or `--guardedness` on again after `--safe` from 1d0ed110e29b72e8aad2d2c9c156aed59c6f3540 would be good to retain (but now in simpler form! :+1: ).

gallais committed:
Merge branch 'master' into Issue3564

gallais committed:
[ manual ] we are now more tolerant

fredrikNordvallForsberg commented on doc/user-manual/language/safe-agda.lagda.rst:
```diff
@@ -51,22 +51,25 @@ then all its imported modules must also be declared safe.
-   (but not both, see above) can be turned on again by giving either
-   ``--guardedness`` or ``--sized-types`` *after* the  ``--safe`` option, e.g.
+   The ``--guardedness`` and ``--sized-types`` options are both on by default.
+   Setting the ``--safe`` option will ignore these defaults and only work if at most
```
"and only work" could perhaps give the wrong impression, how about something like:
```
The ``--guardedness`` and ``--sized-types`` options are both on by default. 
However setting the ``--safe`` option will make Agda turn them off, unless they
are explicitly turned on. Turning on more than one of ``--guardedness`` and
``--sized-types`` together with ``-safe`` results in an error, since these options
are currently inconsistent.

fredrikNordvallForsberg commented on doc/user-manual/language/safe-agda.lagda.rst:
```diff
@@ -51,22 +51,25 @@ then all its imported modules must also be declared safe.
-  will turn on ``--safe``, ``--no-sized-types`` (implied by ``--safe``) and
-  ``--no-guardedness`` (implied by ``--safe``; the earlier option
-  ``--guardedness`` has been overruled by ``--safe``).
+  will turn on ``guardedness``, then ``--safe`` and ``--no-sized-types`` (implied by ``--safe``).
```
Isn't it more useful to point out that this gives the same result? E.g.
```
Hence

    .. code-block:: agda

      {-# OPTIONS --safe --guardedness #-}

will turn on ``--safe``, ``--no-sized-types`` (implied by ``--safe``), and ``--no-guardedness``
(implied by ``--safe``), while either of

    .. code-block:: agda

      {-# OPTIONS --safe #-}

or

    .. code-block:: agda

      {-# OPTIONS --guardedness --safe #-}

will turn on ``--safe``, ``--no-sized-types`` (implied by ``--safe``), and `--guardedness``. Finally e.g.

    .. code-block:: agda

      {-# OPTIONS --guardedness --safe --sized-types #-}

will fail because both ``--guardedness`` and ``--sized-types`` are set together with ``--safe``.
```

gallais replied:
The first one does result in `--guardedness` being turned on because that is
what the user requested.

The second one will result in `--no-guardedness`.

fredrikNordvallForsberg replied:
Sorry, I messed up my copy-pasting! I meant block to be `--safe`, the second to be `--safe --guardedness`, and the third ``--guardedness --safe`` (as it was). But I see you already implemented exactly this, thanks!

gallais committed:
[ manual ] taking @fredrikNordvallForsberg's comment into account

[ci skip]

gallais committed:
[ fix manual ] actually looking at the html output this time

[ci skip]
